src/lightning/app/core/plugin.py

```diff
@@ -103,54 +69,101 @@ def _run_plugin(run: _Run) -> None:
             app=app,
             entrypoint=entrypoint_file,
             start_server=True,
-            env_vars={},
+            env_vars=env_vars if env_vars is not None else {},
             secrets={},
             run_app_comment_commands=True,
         )
         # Used to indicate Lightning has been dispatched
         os.environ["LIGHTNING_DISPATCHED"] = "1"
 
+        runtime.cloudspace_dispatch(
+            project_id=self.project_id,
+            cloudspace_id=self.cloudspace_id,
+            name=name,
+            cluster_id=self.cluster_id,
+        )
+
+    def _setup(
+        self,
+        project_id: str,
+        cloudspace_id: str,
+        cluster_id: str,
+    ) -> None:
+        self.project_id = project_id
+        self.cloudspace_id = cloudspace_id
+        self.cluster_id = cluster_id
+
+
+class _Run(BaseModel):
+    plugin_entrypoint: str
+    source_code_url: str
+    project_id: str
+    cloudspace_id: str
+    cluster_id: str
+    plugin_arguments: Dict[str, str]
+
+
+def _run_plugin(run: _Run) -> List:
+    """Create a run with the given name and entrypoint under the cloudspace with the given ID."""
+    with tempfile.TemporaryDirectory() as tmpdir:
+        download_path = os.path.join(tmpdir, "source.tar.gz")
+        source_path = os.path.join(tmpdir, "source")
+        os.makedirs(source_path)
+
+        # Download the tarball
         try:
-            runtime.cloudspace_dispatch(
-                project_id=run.project_id,
-                cloudspace_id=run.cloudspace_id,
-                name=run.name,
-                cluster_id=run.cluster_id,
-            )
+            # Sometimes the URL gets encoded, so we parse it here
+            source_code_url = urlparse(run.source_code_url).geturl()
+
+            response = requests.get(source_code_url)
+
+            with open(download_path, "wb") as f:
+                f.write(response.content)
         except Exception as e:
```

Comment:
Likely not all exceptions here...

Reply:
Needs to be all in this case as it's a web server, so any uncaught exceptions will give very unhelpful errors 😃